iPad Splitview: Me

frosty opened the pull request:
Introduces `UISplitViewController` to the Me section of the app.

As we've discussed previously, this tries to mimic the system Settings app as much as possible. So the left pane always shows the top level navigation, and the right pane can be used to navigate into each item. I've also made the left pane wider in landscape, to match Settings, and it's always visible in both portrait and landscape.

I've made a number of customisations to `UISplitViewController` to support the behaviour we require, in the subclass `WPSplitViewController`. For example, it handles wrapping and unwrapping child view controllers in navigation controllers where required. 

It also handles automatically populating the right pane with a detail view controller through a protocol `WPSplitViewControllerDetailProvider`. A primary view controller (left pane) can implement the method 

```
func initialDetailViewControllerForSplitView(splitView: WPSplitViewController) -> UIViewController?
```

And return a view controller that should populate the right pane on first appearance. For example, the Me section shows the My Profile detail screen.

To test: 

* Try the Me section out on both iPhone and iPad, moving between orientations and multitasking configurations. 
* Navigate into each item within Me, and try to follow every navigation route to ensure things transition as you'd expect.

Needs review: @kurzee 

We'll review this, and I'll start working on the next section. It's quite likely that `WPSplitViewController` will need some tweaks to support other sections of the app, so we can see where that takes us before we merge any of this.

kurzee commented on WordPress/Classes/Utility/WPImmuTableRows.swift:
```diff
@@ -8,16 +8,18 @@ struct NavigationItemRow : ImmuTableRow {
     let title: String
     let icon: UIImage?
     let action: ImmuTableAction?
+    let showsDisclosureIndicator: Bool
```
Maybe this should allow all accessories via something like `accessoryType`. (Not sure if that is the intended use of `WPImmuTableRows`.

frosty replied:
I like that idea! I've changed it to take an accessory type, but default to `.DisclosureIndicator`.

kurzee commented on WordPress/Classes/ViewRelated/Me/MeViewController.swift:
```diff
@@ -74,7 +84,17 @@ class MeViewController: UITableViewController, UIViewControllerRestoration {
+        }
+
         handler.viewModel = tableViewModel(loggedIn, helpshiftBadgeCount: badgeCount)
+        tableView.selectRowAtIndexPath(selectedIndexPath, animated: false, scrollPosition: .None)
```
Should we maybe only call this code when a `UISplitViewController` is actually initialized?

This might read a bit confusing if someone comes across needing to maintain the code, and also understanding when or why there may be a split view, or if they still need to call `selectRowAtIndexPath`.

frosty replied:
I can't see an easy way around this. After this PR, `MeViewController` is _always_ in a split view, it's just that it may be collapsed at times. I've added extra comments to this section of code to clarify why we're reselecting the selected index path after reloading the table mode.

kurzee commented on WordPress/Classes/ViewRelated/System/WPSplitViewController.swift:
```diff
@@ -0,0 +1,163 @@
+        } else {
+            return UINavigationController(rootViewController: detailViewController)
+        }
+    }
```
Maybe both `initialDetailViewControllerForPrimaryViewController`  and `showDetailViewController` should share the wrapping of navigation controllers, to avoid maintaining this particular step in two spots, but it is pretty simple. 🤔

frosty replied:
I've split the wrapping in a navigation controller into a small helper method, but I can't see a way around both `initialDetailViewControllerForPrimaryViewController` and `showDetailViewController` needing to do this – the problem is that they both have to set the detail view controllers in different ways. `initialDetail...` has to set the split view's `viewController`'s collection to initialise it (it can't call `showDetailViewController`), and then `showDetailViewController` is generally used from that point on.

kurzee commented on WordPress/Classes/ViewRelated/System/WPSplitViewController.swift:
```diff
@@ -0,0 +1,163 @@
+
+extension UIViewController {
+    /// Convenience method
+    var splitViewControllerIsCollapsed: Bool {
```
As similarly noted above, maybe we could change this to read a bit more clearly as to the states it can reveal; perhaps multiple states as something like (roughly): `SplitViewFull`, `SplitViewCollapsed`, `SplitViewNone`.

frosty replied:
I've actually moved things around so this helper is no longer required (for now, anyway).

kurzee commented:
@frosty this looks super good!!

Just those few nit-picky comments. I'll take a second look after the 6+ tweaks are pushed. After that, I'd say it's ready whenever you're ready to 📦 & 🚢!

frosty commented:
@kurzee Thanks for the review and useful comments! I've addressed them where I can (and also fixed the 6+ issue) – do you want to take a second look? I'm going to move on and address another section (branching off this PR) before merging this, just so we can see if this general approach will work okay in other situations.

kurzee commented:
@frosty looks like we lost cell deselection on iPhone and iPad multitasking states. Other than that, 👍 to branch off.

frosty commented:
Closing this now, and we'll deal with it all in #5802. Thanks @kurzee!
